rust/src/asn1/parse_rules.rs

```diff
@@ -0,0 +1,336 @@
+
+/// Parse the asn1 keyword and return a pointer to a `DetectAsn1Data`
+/// containing the parsed options, returns null on failure
+///
+/// # Safety
+///
+/// pointer must be free'd using `rs_detect_asn1_free`
+#[no_mangle]
+pub unsafe extern "C" fn rs_detect_asn1_parse(input: *const c_char) -> *mut DetectAsn1Data {
+    if input.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let arg = match CStr::from_ptr(input).to_str() {
+        Ok(arg) => arg,
+        _ => {
+            return std::ptr::null_mut();
+        }
+    };
+
+    match asn1_parse_rule(&arg) {
+        Ok((_rest, data)) => {
+            let mut data = data;
+
+            // Get configuration value
+            if let Some(max_frames) = crate::conf::conf_get("asn1-max-frames") {
+                if let Ok(v) = max_frames.parse::<u16>() {
+                    data.max_frames = v;
+                } else {
+                    SCLogError!("Could not parse asn1-max-frames: {}", max_frames);
+                    return std::ptr::null_mut();
+                };
+            }
+
+            Box::into_raw(Box::new(data))
+        }
+        Err(e) => {
+            SCLogError!("Malformed asn1 argument: {}", e.to_string());
+            std::ptr::null_mut()
+        }
+    }
+}
+
+/// Free a `DetectAsn1Data` object allocated by Rust
+///
+/// # Safety
+///
+/// ptr must be a valid object obtained using `rs_detect_asn1_parse`
+#[no_mangle]
+pub unsafe extern "C" fn rs_detect_asn1_free(ptr: *mut DetectAsn1Data) {
+    if ptr.is_null() {
+        return;
+    }
+    drop(Box::from_raw(ptr));
+}
+
+/// Struct to hold parsed asn1 keyword options
+#[derive(Debug, PartialEq)]
+pub struct DetectAsn1Data {
+    pub bitstring_overflow: bool,
+    pub double_overflow: bool,
+    pub oversize_length: Option<u32>,
+    pub absolute_offset: Option<u16>,
+    pub relative_offset: Option<i32>,
+    pub max_frames: u16,
+}
+
+impl Default for DetectAsn1Data {
+    fn default() -> DetectAsn1Data {
+        DetectAsn1Data {
+            bitstring_overflow: false,
+            double_overflow: false,
+            oversize_length: None,
+            absolute_offset: None,
+            relative_offset: None,
+            max_frames: ASN1_DEFAULT_MAX_FRAMES,
+        }
+    }
+}
+
+fn parse_u32_number(input: &str) -> IResult<&str, u32> {
+    map_res(digit1, |digits: &str| digits.parse::<u32>())(input)
+}
+
+fn parse_u16_number(input: &str) -> IResult<&str, u16> {
+    map_res(digit1, |digits: &str| digits.parse::<u16>())(input)
+}
+
+fn parse_i32_number(input: &str) -> IResult<&str, i32> {
+    map_res(
+        tuple((opt(tag("-")), digit1)),
+        |(negate, digits): (_, &str)| {
+            digits.parse::<i32>().map(|d| {
+                d * {
+                    if negate.is_some() {
+                        -1
+                    } else {
+                        1
+                    }
+                }
+            })
+        },
+    )(input)
```

Comment:
this looks correct, though I believe using a more imperative style may be easier to understand (approximate code):
```rust
let (negate, digit) = (opt(tag("-")), digit1)(input)?;
let n = if negate.is_some() { -1 } else { 1 };
let (_, digits) = digits.parse::<i32>().map(|d| {
                d *n
})?;
```

Reply:
The `ParseIntError` was previously caught by the `map_res`. You could add `map_res`.
For ex (only compile-tested):
```rust
fn parse_i32_number(input: &str) -> IResult<&str, i32> {
    let (i, negate) = opt(tag("-"))(input)?;
    let (rest, d) = map_res(digit1, |s: &str| s.parse::<i32>())(i)?;
    let n = if negate.is_some() { -1 } else { 1 };
    Ok((rest, d * n))
}
```